File: 2_ActivitySelectionProblem.cpp
#include <iostream>

using namespace std;

class ActivitySelection
{
    struct Activity
    {
        int start;
        int finish;
    };

    Activity act[100];

    int n;

public:

    // Input function
    void getData()
    {
        cout << "Enter number of activities: ";
        cin >> n;

        cout << "\nEnter start and finish times:\n";

        for(int i = 0; i < n; i++)
        {
            cout << "\nActivity " << i + 1 << ":\n";

            cout << "Start time: ";
            cin >> act[i].start;

            cout << "Finish time: ";
            cin >> act[i].finish;
        }
    }

    // Sort activities by finish time
    void sortActivities()
    {
        for(int i = 0; i < n - 1; i++)
        {
            for(int j = i + 1; j < n; j++)
            {
                if(act[i].finish > act[j].finish)
                {
                    swap(act[i], act[j]);
                }
            }
        }
    }

    // Activity Selection
    void selectActivities()
    {
        sortActivities();

        cout << "\nSelected Activities:\n";

        int i = 0;

        cout << "Activity 1 "
             << "(" << act[i].start
             << "," << act[i].finish << ")\n";

        // Select remaining activities
        for(int j = 1; j < n; j++)
        {
            if(act[j].start >= act[i].finish)
            {
                cout << "Activity "
                     << j + 1
                     << " ("
                     << act[j].start
                     << ","
                     << act[j].finish
                     << ")\n";

                i = j;
            }
        }
    }
};

int main()
{
    ActivitySelection a;

    a.getData();

    a.selectActivities();

    return 0;
}

File: 2_ActivitySelectionProblem.java
import java.util.Scanner;

public class ActivitySelection {
    private static class Activity {
        int start;
        int finish;

        Activity(int start, int finish) {
            this.start = start;
            this.finish = finish;
        }
    }

    private Activity[] act = new Activity[100];
    private int n;
    private Scanner sc = new Scanner(System.in);

    public void getData() {
        System.out.print("Enter number of activities: ");
        n = sc.nextInt();

        System.out.println("\nEnter start and finish times:");

        for (int i = 0; i < n; i++) {
            System.out.println("\nActivity " + (i + 1) + ":");
            System.out.print("Start time: ");
            int start = sc.nextInt();
            System.out.print("Finish time: ");
            int finish = sc.nextInt();
            act[i] = new Activity(start, finish);
        }
    }

    public void sortActivities() {
        for (int i = 0; i < n - 1; i++) {
            for (int j = i + 1; j < n; j++) {
                if (act[i].finish > act[j].finish) {
                    Activity temp = act[i];
                    act[i] = act[j];
                    act[j] = temp;
                }
            }
        }
    }

    public void selectActivities() {
        sortActivities();

        System.out.println("\nSelected Activities:");

        int i = 0;
        System.out.println("Activity 1 (" + act[i].start + "," + act[i].finish + ")");

        for (int j = 1; j < n; j++) {
            if (act[j].start >= act[i].finish) {
                System.out.println("Activity " + (j + 1) + " (" + act[j].start + "," + act[j].finish + ")");
                i = j;
            }
        }
    }

    public static void main(String[] args) {
        ActivitySelection a = new ActivitySelection();
        a.getData();
        a.selectActivities();
    }
}
